### PR TITLE
pool/SNOW-902608 #2/4 new pool version v2

### PR DESCRIPTION
**Pull request by sfc-gh-mhofman**
### Description
Connection Pool v2
- introduction of connection manager handling multiple pool of sessions
- ability to switch Connection Manager between v1 and v2
- controlling pool settings for each pool independently
- default pool version: v1

### Checklist
- [x] Code compiles correctly
- [x] Code is formatted according to [Coding Conventions](../CodingConventions.md)
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing (`dotnet test`)
- [x] Extended the README / documentation, if necessary
- [x] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: Snowflake.Data/Client/SnowflakeDbConnectionPool.cs
Comment (sfc-gh-mhofman) [inline Snowflake.Data/Client/SnowflakeDbConnectionPool.cs:35]: i'll remove this {connectionString}

File: Snowflake.Data/Core/Session/ConnectionPoolManager.cs
Comment (sfc-gh-mhofman) [inline Snowflake.Data/Core/Session/ConnectionPoolManager.cs:32]: i'll remove {connectionString}

File: Snowflake.Data/Core/Session/ConnectionPoolManager.cs
Comment (sfc-gh-mhofman) [inline Snowflake.Data/Core/Session/ConnectionPoolManager.cs:44]: to be removed

File: Snowflake.Data/Core/Session/SessionPool.cs
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SessionPool.cs:30]: Good idea, will do

Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SessionPool.cs:30]: On second thought, it would be better but requires passing the same factory through multiple layers of code. Is it really worth? From test perspective it will be just injected into another class using some static method. Same as now.

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs:39]: SessionFactory is internal so should be s_.
I think we should rather have a method to change SessionPool attributes.

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs:22]: const?

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs
Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs:39]: I'm fixing naming convention.

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs
Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs:57]: It's testing that same pool is returned when called multiple times.

File: Snowflake.Data.Tests/Util/PoolConfig.cs
Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/Util/PoolConfig.cs:30]: You're right, so these should go after

File: Snowflake.Data/Core/Session/SessionPool.cs
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SessionPool.cs:86]: my bad, removing

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs:49]: I feel a little bit uncomfortable seeing getting a private field: 
```
sessionPool._connectionString
```

maybe for these fileds (connectionString and password) could be created internal properties so here we could reference internal property instead of private field?

File: Snowflake.Data/Core/Session/ConnectionCacheManager.cs
Comment (sfc-gh-knozderko) [reply Snowflake.Data/Core/Session/ConnectionCacheManager.cs:2]: what about it?

File: Snowflake.Data/Core/Session/SFSession.cs
Comment (sfc-gh-knozderko) [reply Snowflake.Data/Core/Session/SFSession.cs:71]: what about it?

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs
Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs:49]: They are internal

**Comment by sfc-gh-mhofman** [comment]
The code is inactive and there is no public api to enable the new pool for now. 
Hence, once approved we can merge and wait for security review to be completed in the background.

File: Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolIT.cs
Comment (sfc-gh-dstempniak) [reply Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolIT.cs:236]: will you?

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs
Comment (sfc-gh-dstempniak) [reply Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerTest.cs:49]: maybe make it internal property ConnectionString with get only?

File: Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerSwitchTest.cs
Comment (sfc-gh-dstempniak) [inline Snowflake.Data.Tests/UnitTests/ConnectionPoolManagerSwitchTest.cs:16]: As we test here SnowflakeDbConnectionPool, maybe the class should be called SnowflakeDbConnectionPoolTest?
And later we could add here other unit tests of SnowflakeDbConnectionPool.
Just an idea.

File: Snowflake.Data/Core/Session/SFSession.cs
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SFSession.cs:71]: Auto property is a better thing - similarily to what we ended up with in SessionPool

**Review by sfc-gh-dstempniak — APPROVED** [review]
LGTM

**Review by sfc-gh-knozderko — APPROVED** [review]

**Pull request merged**
